feat: convert to grayscale

higor-melo opened the pull request:
Add ability to convert the image to grayscale but keep the frame mode as the desired one from the calibration parameters.
[sc-75363]

higor-melo committed:
feat: allow maintaining the desired frame mode on a grayscaled image

higor-melo committed:
feat: add unit tests

We compare if each of the pixels have the same RGB color values
for each pixel, therefore confirming the image is grayscale.

higor-melo committed:
fix: do not modify the input frame

higor-melo committed:
fix: black image if frame_time is missing

updated time from sample

chhtz commented on tasks/MonoTask.cpp:
```diff
@@ -1,22 +1,22 @@
 /* Generated from orogen/lib/orogen/templates/tasks/Task.cpp */
 
 #include "MonoTask.hpp"
+#include <base-logging/Logging.hpp>
```
Do you use logging anywhere in the code?

higor-melo replied:
While writing the tests I had some issues and i used logging to pinpoint the problems.
Now it's useless, i'll remove it.

chhtz commented on image_preprocessing.orogen:
```diff
@@ -1,3 +1,5 @@
+# frozen_string_literal: true
+
```
What are these comments for?

higor-melo replied:
This forbids allocation of new memory to the same string and is implemented by default on Ruby 3.0. However, we're currently running an older version of Ruby 2.7 and my linter placed the `#frozen_string_literal` comment by default.

chhtz commented on image_preprocessing.orogen:
```diff
@@ -9,40 +11,52 @@ using_library "frame_helper"
 
-   property("offset_y",   "int", 0).
-       doc "the offset to be used on the left margin in case the of resizing widescreen images"
+    # the offset to be used on the left margin in case the of resizing widescreen
```
'in case **~the~** of resizing'

chhtz commented on image_preprocessing.orogen:
```diff
@@ -9,40 +11,52 @@ using_library "frame_helper"
+    # images"
+    property("offset_x",   "int", 0)
+
+    # the offset to be used on the top margin in case the of resizing widescreen
```
'in case **~the~** of resizing'

chhtz commented on image_preprocessing.orogen:
```diff
@@ -9,40 +11,52 @@ using_library "frame_helper"
-    input_port("color_frame", ro_ptr("/base/samples/frame/Frame")).
-        doc("Color Frame for DepthImage if availible")
+    input_port("color_frame", ro_ptr("/base/samples/frame/Frame"))
+        .doc("Color Frame for DepthImage if availible")
```
avail_**a**_ble

chhtz commented on manifest.xml:
```diff
@@ -10,6 +10,9 @@
   <depend package="drivers/orogen/transformer"/>
   <depend package="image_processing/frame_helper"/>
   <depend package="drivers/orogen/aggregator" />
+  <depend package="opencv" />
```
I think opencv is only used indirectly via the FrameHelper, I would therefore not add a direct dependency here.

chhtz commented on image_preprocessing.orogen:
```diff
@@ -9,40 +11,52 @@ using_library "frame_helper"
-   property("frame_mode", "/base/samples/frame/frame_mode_t", :MODE_BAYER).
-       doc "output frame mode"
+    # The mode of the output image
+    property("frame_mode", "/base/samples/frame/frame_mode_t", :MODE_BAYER)
```
I assume keeping the `.doc` string is advantageous, e.g., when config-yaml files are created (but I did not check if comments are processed as well).
Same for the next doc changes

chhtz commented on manifest.xml:
```diff
@@ -10,6 +10,9 @@
+  <depend package="opencv" />
+
+  <test_depend name="tools/syskit" />
   <versioncontrol type="dummy-version-control" url="dummy-version-control-url" />
```
We should probably also fix these "dummy" entries here (or remove them ...)

higor-melo replied:
I'm leaning towards removing them, what do you think?

chhtz reviewed (changes requested):
Generally, I approve. Some open points:
* You add the config to the `BaseTask` but only the `MonoTask` actually implements the functionality
* This may make minor difference: In case resizing/undistorting happens, should this happen simultaneously to the grayscale conversion, and only afterwards convert the grayscale frame back to a color frame?

higor-melo commented:
> Generally, I approve. Some open points:
> 
> * You add the config to the `BaseTask` but only the `MonoTask` actually implements the functionality
> * This may make minor difference: In case resizing/undistorting happens, should this happen simultaneously to the grayscale conversion, and only afterwards convert the grayscale frame back to a color frame?

IMO it's better to leave it with the `MonoTask` to avoid having 2 conversions between different frame modes. For example, if one was to input a frame `MODE_BGR` and expect a `MODE_RGB` however this frame was also to be actually a grayscale frame. If i were to implement the conversion on the BaseTask to avoid modifying the frame_mode I would do the following:
BGR -> GRAYSCALE -> BGR -> RGB

Instead of packing the code on the MonoTask which also handles the frame_mode conversion and do:
BGR -> GRAYSCALE -> RGB

jhonasiv commented:
> > Generally, I approve. Some open points:
> > 
> > * You add the config to the `BaseTask` but only the `MonoTask` actually implements the functionality
> > * This may make minor difference: In case resizing/undistorting happens, should this happen simultaneously to the grayscale conversion, and only afterwards convert the grayscale frame back to a color frame?
> 
> IMO it's better to leave it with the `MonoTask` to avoid having 2 conversions between different frame modes. For example, if one was to input a frame `MODE_BGR` and expect a `MODE_RGB` however this frame was also to be actually a grayscale frame. If i were to implement the conversion on the BaseTask to avoid modifying the frame_mode I would do the following: BGR -> GRAYSCALE -> BGR -> RGB
> 
> Instead of packing the code on the MonoTask which also handles the frame_mode conversion and do: BGR -> GRAYSCALE -> RGB

His point for **1** is that the config should be defined in MonoTask, as its the only one that actually has the functionality.

higor-melo committed:
fix: convert_to_grayscale is only used on MonoTask

Also reverted the documentation changes back to .doc

higor-melo committed:
chore: remove unused include and depend package

OpenCv is only used indirectly by the FrameHelper.
Removed versioncontrol type.

higor-melo committed:
fix: remove frozen_string_literal from orogen

Also linting

chhtz commented on tasks/MonoTask.cpp:
```diff
@@ -33,6 +33,7 @@ bool MonoTask::startHook()
         return false;
 
     frame_helper.setCalibrationParameter(_calibration.value());
+    m_convert_to_grayscale = _convert_to_grayscale.value();
```
This should be in the `configureHook()` (keeping it in the `updateHook()` is ok as well, since it is just checking a flag).

higor-melo replied:
Got it, to align it with the other tasks I'll be keeping it on the start hook

chhtz commented:
Besides my last change request (`configureHook` vs `startHook`) I'm fine with merging this.

chhtz reviewed (approved):
I'll merge this tomorrow morning, unless objections arise until then.

higor-melo commented:
> I'll merge this tomorrow morning, unless objections arise until then.

Thank you!
